### PR TITLE
fix: Eliminate global var from protocol response test generation

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
https://github.com/awslabs/aws-sdk-swift/issues/1612

## Description of changes
Eliminates a global variable that was used during protocol response test generation to track shapes that had already had Equatable conformance rendered.

The functionality has been removed from the protocol test generator, which runs in parallel with other rendering activities, and is now split out to two places during preprocessing:
- The new `TestEquatableConformanceTransformer` model transformer now marks shapes that need Equatable for protocol response test verification with a new synthetic trait named `TestEquatableConformanceTrait`.
- Actual rendering of Equatable conformance for protocol tests now happens in a new Swift integration named `TestEquatableConformanceIntegration`.

## Scope
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/customtraits/TestEquatableConformanceTrait.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package software.amazon.smithy.swift.codegen.customtraits
+
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.traits.Trait
+
+class TestEquatableConformanceTrait : Trait {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/customtraits/TestEquatableConformanceTrait.kt:7]: This trait is applied to shapes that need to have Equatable conformance added to them for protocol tests.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt [hunk 2]
```diff
@@ -119,82 +109,12 @@ open class HttpProtocolUnitTestResponseGenerator protected constructor(builder:
...
-    }
-
     open class Builder : HttpProtocolUnitTestGenerator.Builder<HttpResponseTestCase>() {
         override fun build(): HttpProtocolUnitTestGenerator<HttpResponseTestCase> {
             return HttpProtocolUnitTestResponseGenerator(this)
         }
     }
 }
-
-val hasBeenRenderedEquatable = mutableSetOf<String>()
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt:200]: This is the global var that has been causing the intermittent failures.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt [hunk 3]
```diff
@@ -119,82 +109,12 @@ open class HttpProtocolUnitTestResponseGenerator protected constructor(builder:
...
 
     protected open fun renderAssertions(test: HttpResponseTestCase, outputShape: Shape) {
-        val nestedShapes = model.getNestedShapes(operation)
-        for (shape in nestedShapes) {
-            when (shape.type) {
-                ShapeType.STRUCTURE -> renderEquatable(shape)
-                ShapeType.UNION -> renderEquatable(shape)
-                else -> {}
-            }
-        }
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt:129]: Getting nested shapes for the test's response or error, and marking them so they get conformance added to them later, is now taken care of with a model transformation - see `EquatableConformanceTransformer` below.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt [hunk 4]
```diff
@@ -119,82 +109,12 @@ open class HttpProtocolUnitTestResponseGenerator protected constructor(builder:
...
         writer.write("XCTAssertEqual(actual, expected)")
     }
 
-    private fun identifier(ctx: ProtocolGenerator.GenerationContext, shape: Shape): String {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt:133]: All of this rendering code is moved to the SwiftIntegration named `TestEquatableConformanceIntegration` below.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/swiftintegrations/TestEquatableConformanceIntegration.kt [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import software.amazon.smithy.swift.codegen.SwiftDependency
+import software.amazon.smithy.swift.codegen.core.SwiftCodegenContext
+import software.amazon.smithy.swift.codegen.customtraits.EquatableConformanceTrait
+import software.amazon.smithy.swift.codegen.customtraits.TestEquatableConformanceTrait
+import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
+import software.amazon.smithy.swift.codegen.integration.SwiftIntegration
+import software.amazon.smithy.swift.codegen.model.hasTrait
+import software.amazon.smithy.swift.codegen.swiftmodules.SwiftTypes
+
+class TestEquatableConformanceIntegration : SwiftIntegration {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/swiftintegrations/TestEquatableConformanceIntegration.kt:20]: This SwiftIntegration does the code generation of Equatable conformance for types that need Equatable for protocol response tests, but don't already have it auto-rendered in the SDK.

The rendering code is copied straight out of the `HttpProtocolUnitTestResponseGenerator`.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/TestEquatableConformanceTransformer.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.shapes.ShapeType
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.transform.ModelTransformer
+import software.amazon.smithy.protocoltests.traits.HttpResponseTestsTrait
+import software.amazon.smithy.swift.codegen.customtraits.TestEquatableConformanceTrait
+
+object TestEquatableConformanceTransformer {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/TestEquatableConformanceTransformer.kt:13]: A new model transformer that applies the `TestEquatableConformanceTrait` to any shape that is used in a response protocol test, or any shape nested within such a shape.

See inline comments for greater detail.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftCodegenPlugin.kt [hunk 7]
```diff
@@ -56,6 +57,7 @@ class SwiftCodegenPlugin : SmithyBuildPlugin {
             resolvedModel = NestedShapeTransformer.transform(resolvedModel, settings.getService(resolvedModel))
             resolvedModel = UnionIndirectivizer.transform(resolvedModel)
             resolvedModel = EquatableConformanceTransformer.transform(resolvedModel, settings.getService(resolvedModel))
+            resolvedModel = TestEquatableConformanceTransformer.transform(resolvedModel, settings.getService(resolvedModel))
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/SwiftCodegenPlugin.kt:60]: Performs the new model transform during codegen, along with the other model transforms.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/ShapeExt.kt [hunk 8]
```diff
@@ -123,15 +123,17 @@ fun Model.getNestedShapes(serviceShape: ServiceShape): Set<Shape> {
...
         .select(this)
+        .map { it as StructureShape }
+        .toSet()
 }
 
-fun Model.getNestedShapes(shape: Shape): Set<Shape> {
+fun Model.getNestedShapes(memberShape: MemberShape): Set<Shape> {
     return Selector
-        .parse("operation [id='${shape.id}'] ~> *")
+        .parse("member [id='${memberShape.id}'] ~> *")
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/model/ShapeExt.kt:136]: The diff here is confusing, but I deleted
`fun Model.getNestedShapes(shape: Shape): Set<Shape>` and replaced it with
`fun Model.getNestedErrors(serviceShape: ServiceShape): Set<StructureShape>`.

**Review by dayaffe — APPROVED** [review]

**Pull request merged**